### PR TITLE
[GRAPH] Use channel shuffling only for IB systems

### PR DESCRIPTION
**Pull request by nileshnegi**
## Details

**Work item:**
Internal

**What were the changes?**  
- Re-introduce channel shuffling for MI300x systems with Infiniband interconnect.
- Fix channels=48 for RoCE systems 

**Why were the changes made?**  
Channel shuffling improves performance on IB systems, but adversely affects RoCE systems.

**How was the outcome achieved?**  
`NCCL_IB_GID_INDEX` needs to be defined as **3** for RoCE systems. If this is the case, do not use channel shuffling rings and use channels=48, else, use channel shuffling and channels=64.

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

**Review by gilbertlee-amd — APPROVED** [review]

File: src/graph/rome_models.cc [hunk 1]
```diff
@@ -642,8 +889,9 @@ static struct rcclRomeModel rome_model_81 = {
                 "N6 6 2 0 7 5 3 1 4 N4|"
                 "N7 7 3 2 6 0 4 1 5 N5|"
                 "N1 1 0 2 4 3 5 7 6 N6|"
-                "N0 0 1 2 3 4 5 6 7 N7",
-  .options    = "noCpuCheck=1,tuning=5,disableNumaMatching=1",
+		"N0 0 1 2 3 4 5 6 7 N7",
+
+  .options    = "noCpuCheck=1,tuning=5,disableNumaMatching=1,gid3=1",
```
Comment (gilbertlee-amd) [inline src/graph/rome_models.cc]: For code clarity, I think we should rename this option to isRoCE (or however the capitalization should be).

Comment (gilbertlee-amd) [reply src/graph/rome_models.cc]: Alternatively, set this to gidIndex=3 and check that we match that value if the option is set.

**Review by gilbertlee-amd — APPROVED** [review]

**Review by wenkaidu — APPROVED** [review]

File: src/graph/rome_models.cc [hunk 2]
```diff
@@ -1367,12 +1616,23 @@ ncclResult_t parseRome4P2H(struct ncclTopoSystem* system, struct ncclTopoGraph*
...
   for (i = 0; i < sizeof(romeTopoModels)/sizeof(romeTopoModels[0]); i++) {
     bool ignore_cpu = checkOption(romeTopoModels[i].options, "noCpuCheck");
     if (!ignore_cpu && (arch != NCCL_TOPO_CPU_ARCH_X86 || vendor != NCCL_TOPO_CPU_VENDOR_AMD || model != NCCL_TOPO_CPU_TYPE_ROME))
       continue;
+
     bool ignore_numa = checkOption(romeTopoModels[i].options, "disableNumaMatching");
     if (!ignore_numa && romeTopo.nCpus != romeTopoModels[i].nCpus) continue;
+
+    // check if isRoCE=1 is defined in model struct
+    bool optionsIsRoCE = checkOption(romeTopoModels[i].options, "isRoCE");
```
Comment (wenkaidu) [inline src/graph/rome_models.cc]: what happen with models doesn't have "isRoCE" option?

Comment (nileshnegi) [reply src/graph/rome_models.cc]: `checkOption` returns false if "isRoCE" is not found in options, so `optionsIsRoCE` will be 0.\
variable `isRoCE` in line 1621 is initialized as false, so line 1634: `if (optionsIsRoCE != isRoCE) continue;` will not come into effect.

verified with model match for single node run, which shows "Rome model index 38 with GPU mapping: 0 1 2 3 4 5 6 7" i.e. `rome_model_79` which does not have "isRoCE" in options

Comment (wenkaidu) [reply src/graph/rome_models.cc]: if the purpose is to skip "if (optionsIsRoCE != isRoCE) continue" in this case, optionsIsRoCE needs to have 3 values: not present, true and false.

**Review by wenkaidu — APPROVED** [review]

File: src/graph/connect.cc [hunk 3]
```diff
@@ -627,8 +627,19 @@ ncclResult_t ncclTopoPostset(struct ncclComm* comm, int* firstRanks, int* treePa
   // Only use full MAXCHANNELS for gfx94x
   int maxChannels = IsArchMatch(comm->topo->nodes[GPU].nodes[0].gpu.gcn, "gfx94") ? MAXCHANNELS : 2*CHANNEL_LIMIT;
 
+  // Define channels=64 for multi-node systems
+  int multiNodeNchannels = 64;
```
Comment (wenkaidu) [inline src/graph/connect.cc]: We should only use more than 32 channels on MI300

**Review by wenkaidu — APPROVED** [review]

**Review by wenkaidu — APPROVED** [review]

**Pull request merged**